lib/ansible/galaxy/dependency_resolution/gpg.py

```diff
@@ -0,0 +1,298 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2022, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""Signature verification helpers."""
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.errors import AnsibleError
+from ansible.module_utils.common.text.converters import to_text
+from ansible.module_utils.urls import open_url
+from ansible.utils.display import Display
+
+import os
+import subprocess
+import textwrap
+
+from dataclasses import dataclass, fields as dc_fields
+from urllib.error import HTTPError
+
+display = Display()
+
+
+def run_gpg_verify(b_manifest_file, signature, keyring):
+    status_fd_read, status_fd_write = os.pipe()
+
+    cmd = [
+        'gpg',
+        f'--status-fd={status_fd_write}',
+        '--verify',
+        '--batch',
+        '--no-tty',
+        '--no-default-keyring',
+        f'--keyring={keyring}',
+        '-',
+        b_manifest_file,
+    ]
+
+    try:
+        p = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            pass_fds=(status_fd_write,),
+        )
+        stdout, stderr = p.communicate(input=signature)
+    except subprocess.SubprocessError as err:
+        cmd_str = ' '.join([to_text(arg) for arg in cmd])
+        display.error(f"Failed during GnuPG verification with command '{cmd_str}':\n{err}")
+        p = None
+    finally:
+        os.close(status_fd_write)
+
+    if p is None:
+        return
+
+    rc = p.returncode
+    with os.fdopen(status_fd_read) as f:
+        return f.read(), p.returncode
+
+
+def parse_gpg_errors(status_out, rc):
+    errors = []
+    for line in status_out.splitlines():
+        if not line:
+            continue
+        try:
+            _dummy, status, remainder = line.split(None, 2)
+        except ValueError:
+            _dummy, status = line.split(None, 1)
+            remainder = None
+        try:
+            cls = GPG_ERROR_MAP[status]
+        except KeyError:
+            continue
+
+        fields = [status]
+        if remainder:
+            fields.extend(
+                remainder.split(
+                    None,
+                    len(dc_fields(cls)) - 2
+                )
+            )
+
+        yield cls(*fields)
+
+
+def get_signature_from_url(url, quiet=False):
+    if not quiet:
+        display.vvvv(f"Using signature at {url}")
+    try:
+        resp = open_url(url, follow_redirects='safe')
```

Comment:
This returns a context manager. Make sure to call `resp.close()` or use a `with`-block. Otherwise, there will be a `ResourceWarning` raised (should be visible with `python -Xdev` or `-Werror`).

Reply:
Fixed.